### ISSUE TITLE
error in custom metric function

### ISSUE THREAD

**Issue by ruddnr** [tidymodels/yardstick#431]
## The problem

I'm having trouble with using custom performance metrics.
I followed the tutorial on the website but I get an error message below. 

> unique notes:
> ─────────────────────────────────────────────────────────────────────────────────────────────────
> Error in `metric_set()`:
> ! Failed to compute `mse()`.
> Caused by error in `UseMethod()`:
> ! no applicable method for 'mse' applied to an object of class "c('tbl_df', 'tbl', 'data.frame')"

## Reproducible example

```r
mse_impl <- function(truth, estimate, case_weights = NULL) {
  mean((truth - estimate) ^ 2)
}

mse_vec <- function(truth, estimate, na_rm = TRUE, case_weights = NULL, ...) {
  check_numeric_metric(truth, estimate, case_weights)

  if (na_rm) {
    result <- yardstick_remove_missing(truth, estimate, case_weights)

    truth <- result$truth
    estimate <- result$estimate
    case_weights <- result$case_weights
  } else if (yardstick_any_missing(truth, estimate, case_weights)) {
    return(NA_real_)
  }

  mse_impl(truth, estimate, case_weights = case_weights)
}

mse <- function(data, ...) {
  UseMethod("mse")
}

mse <- new_numeric_metric(mse, direction = "minimize")

mse.data.frame <- function(data, truth, estimate, na_rm = TRUE, case_weights = NULL, ...) {

  numeric_metric_summarizer(
    name = "mse",
    fn = mse_vec,
    data = data,
    truth = !!enquo(truth),
    estimate = !!enquo(estimate),
    na_rm = na_rm,
    case_weights = !!enquo(case_weights)
  )
}

library(tidymodels)
data(ames)

ames <- mutate(ames, Sale_Price = log10(Sale_Price))

set.seed(502)
ames_split <- initial_split(ames, prop = 0.80, strata = Sale_Price)
ames_train <- training(ames_split)
ames_test  <-  testing(ames_split)

ames_rec <-
  recipe(Sale_Price ~ Neighborhood + Gr_Liv_Area + Year_Built + Bldg_Type +
           Latitude + Longitude, data = ames_train) %>%
  step_log(Gr_Liv_Area, base = 10) %>%
  step_other(Neighborhood, threshold = 0.01) %>%
  step_dummy(all_nominal_predictors()) %>%
  step_interact( ~ Gr_Liv_Area:starts_with("Bldg_Type_") ) %>%
  step_ns(Latitude, Longitude, deg_free = 20)

rf_model <-
  rand_forest(trees = 100) %>%
  set_engine("ranger") %>%
  set_mode("regression")

rf_wflow <-
  workflow() %>%
  add_formula(
    Sale_Price ~ Neighborhood + Gr_Liv_Area + Year_Built + Bldg_Type +
      Latitude + Longitude) %>%
  add_model(rf_model)

set.seed(1001)
ames_folds <- vfold_cv(ames_train, v = 10)

keep_pred <- control_resamples(save_pred = TRUE, save_workflow = TRUE)
model_metrics <- metric_set(mse, rmse)

set.seed(1003)
rf_res <- rf_wflow %>% fit_resamples(resamples = ames_folds, control = keep_pred, metrics = model_metrics)
```

I have no problem with the code below, but it fails if I try to use it in `fit_resamples`.

```r
data("solubility_test")

numeric_mets <- metric_set(mse, rmse)

numeric_mets(solubility_test, solubility, prediction)
```

I tried to include the function in my package, but the same problem happens. 

Thanks in advance.

**Comment by ruddnr** [comment]
Oh, It works after I update the package to the development version, but it still fails if I try to use parallel computing.

```r
all_cores <- parallel::detectCores(logical = FALSE)
cl <- makePSOCKcluster(all_cores)
registerDoParallel(cl)
set.seed(1003)
rf_res <- rf_wflow %>% fit_resamples(resamples = ames_folds, control = keep_pred, metrics = model_metrics)
```

**Comment by EmilHvitfeldt** [comment]
Hello @ruddnr 👋  you need to export certain function to your cluster for it to know that you need them. You can do this with `parallel::clusterExport()`. I'm using `ls(pattern = "mse")` as a shorthand for `c("mse", "mse_impl", "mse_vec", "mse.data.frame")`.

Secondly, you will need to namespace the functions you are using in the functions, like I have done below

``` r
library(yardstick)

mse_impl <- function(truth, estimate, case_weights = NULL) {
  mean((truth - estimate) ^ 2)
}

mse_vec <- function(truth, estimate, na_rm = TRUE, case_weights = NULL, ...) {
  yardstick::check_numeric_metric(truth, estimate, case_weights)
  
  if (na_rm) {
    result <- yardstick::yardstick_remove_missing(truth, estimate, case_weights)
    
    truth <- result$truth
    estimate <- result$estimate
    case_weights <- result$case_weights
  } else if (yardstick::yardstick_any_missing(truth, estimate, case_weights)) {
    return(NA_real_)
  }
  
  mse_impl(truth, estimate, case_weights = case_weights)
}

mse <- function(data, ...) {
  UseMethod("mse")
}

mse <- yardstick::new_numeric_metric(mse, direction = "minimize")

mse.data.frame <- function(data, truth, estimate, na_rm = TRUE, case_weights = NULL, ...) {
  
  yardstick::numeric_metric_summarizer(
    name = "mse",
    fn = mse_vec,
    data = data,
    truth = !!rlang::enquo(truth),
    estimate = !!rlang::enquo(estimate),
    na_rm = na_rm,
    case_weights = !!rlang::enquo(case_weights)
  )
}

library(tidymodels)
data(ames)

ames <- mutate(ames, Sale_Price = log10(Sale_Price))

set.seed(502)
ames_split <- initial_split(ames, prop = 0.80, strata = Sale_Price)
ames_train <- training(ames_split)
ames_test  <-  testing(ames_split)

ames_rec <-
  recipe(Sale_Price ~ Neighborhood + Gr_Liv_Area + Year_Built + Bldg_Type +
           Latitude + Longitude, data = ames_train) %>%
  step_log(Gr_Liv_Area, base = 10) %>%
  step_other(Neighborhood, threshold = 0.01) %>%
  step_dummy(all_nominal_predictors()) %>%
  step_interact( ~ Gr_Liv_Area:starts_with("Bldg_Type_") ) %>%
  step_ns(Latitude, Longitude, deg_free = 20)

rf_model <-
  rand_forest(trees = 100) %>%
  set_engine("ranger") %>%
  set_mode("regression")

rf_wflow <-
  workflow() %>%
  add_formula(
    Sale_Price ~ Neighborhood + Gr_Liv_Area + Year_Built + Bldg_Type +
      Latitude + Longitude) %>%
  add_model(rf_model)

set.seed(1001)
ames_folds <- vfold_cv(ames_train, v = 10)

keep_pred <- control_resamples(save_pred = TRUE, save_workflow = TRUE)
model_metrics <- metric_set(mse, rmse)

library(doParallel)
#> Loading required package: foreach
#> 
#> Attaching package: 'foreach'
#> The following objects are masked from 'package:purrr':
#> 
#>     accumulate, when
#> Loading required package: iterators
#> Loading required package: parallel
all_cores <- parallel::detectCores(logical = FALSE)
cl <- parallel::makePSOCKcluster(all_cores)

parallel::clusterExport(cl, ls(pattern = "mse"))

registerDoParallel(cl)
set.seed(1003)
rf_res <- rf_wflow %>% 
  fit_resamples(
    resamples = ames_folds, 
    control = keep_pred, 
    metrics = model_metrics
  )

stopCluster(cl)

collect_metrics(rf_res)
#> # A tibble: 2 × 6
#>   .metric .estimator    mean     n  std_err .config             
#>   <chr>   <chr>        <dbl> <int>    <dbl> <chr>               
#> 1 mse     standard   0.00534    10 0.000457 Preprocessor1_Model1
#> 2 rmse    standard   0.0725     10 0.00302  Preprocessor1_Model1
```

<sup>Created on 2023-05-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by EmilHvitfeldt** [comment]
related to https://github.com/tidymodels/recipes/issues/1031

**Comment by ruddnr** [comment]
Thank you so much! 
Sorry that I think I made a duplicate question. 
I solved the issue by including my custom function in my own package and adding a namespace to each function.

I didn't know that adding namespace to each function is so crucial when using my function in parallel processing. 

Before I add namespace to each function, I tried using "pkgs" argument in `control_resamples` function but it failed.
but I could solve the issue by sending my whole package to each worker.

```r
all_cores <- parallel::detectCores(logical = FALSE)
cl <- makePSOCKcluster(all_cores)
parallel::clusterEvalQ(cl,library("kwlib"))
registerDoParallel(cl)
```

Thank you so much for your help.
